### PR TITLE
feat: scrape rules ( on ice until #714 spike completion )

### PR DESCRIPTION
**Pull request by aakropotkin**
## Proposed Changes

- Adds the ability to provide `pkgdb` scrape creation rules.
- Rules can be with `--rules FILE`, or by editing a hard coded fallback: `pkgdb/src/pkgdb/rules.json.hh`.
- This PR has some limitations which require further work before we can close the connected issue.
  + see various TODO and FIXME notes added in this PR for specific issues.

Notable limitations:
- Cannot nest `allow*` under `disallowRecursive` ( but docs claim we can ).
  + Complete this functionality.
  + `applyRules()` - supports nesting `allow*` under `disallowRecursive`.
  + `scrape()` - has a FIXME which currently prevents this from working.
- No database invalidation when rules change.
  + It is strongly recommended that some hash ( similar to fingerprint ) be added to a meta fields of DBs to invalidate them.
  + For now, delete your caches during local dev between rule changes.
  + You likely won't want to use `--rules` until you have invalidation completed.
- Needs more tests.
- The pipeline to pass `rulesPath` from the arg parser to `scrape` should not use global variables.
  + Add `ScrapeRulesRaw` or something to `PkgDbInput`.

## Release Notes
- Additional Darwin packages are available.

### CODE REVIEW

File: pkgdb/Makefile [hunk 1]
```diff
@@ -539,6 +539,9 @@ $(_TEST_LOCKFILES): %/manifest.lock: %/manifest.toml bin/pkgdb
 	_PKGDB_GA_REGISTRY_REF_OR_REV="e8039594435c68eb4f780f3e9bf3972a7399c4b1"  \
 	  bin/pkgdb manifest lock --ga-registry --manifest $< > $@;
 
+# TODO: When we can run `pkgdb' in parallel this can be removed.
+.NOTPARALLEL: $(_TEST_LOCKFILES) test-lockfiles test-data
```
Comment (aakropotkin) [inline pkgdb/Makefile:543]: Fixes issue running `make -j tests`

File: pkgdb/tests/scrape-rules.bats [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+	mkdir -p "$BATS_FILE_TMPDIR"
+	# We don't parallelize these to avoid DB sync headaches and to recycle the
+	# cache between tests.
+	# Nonetheless this file makes an effort to avoid depending on past state in
+	# such a way that would make it difficult to eventually parallelize in
+	# the future.
+	export BATS_NO_PARALLELIZE_WITHIN_FILE=true
+
+	# Define rules.json content
+	rules_json='{
```
Comment (aakropotkin) [inline pkgdb/tests/scrape-rules.bats:27]:
```suggestion
       # FIXME: formatting
	rules_json='{
```

File: pkgdb/tests/search.bats [hunk 3]
```diff
@@ -259,7 +281,10 @@ genParamsNixpkgsFlox() {
 
 # Check fallback behavior.
 @test "search-params with empty object" {
-  run $PKGDB_SEARCH_PARAMS_BIN '{}'
+  if [[ -z "${PKGDB_SEARCH_PARAMS_BIN:=$( command -v search-params; )}" ]]; then
```
Comment (aakropotkin) [inline pkgdb/tests/search.bats:284]:
```suggestion
  if [ -z "${PKGDB_SEARCH_PARAMS_BIN:=$( command -v search-params; )}" ]; then
```

File: pkgs/flox-pkgdb/default.nix [hunk 4]
```diff
@@ -174,13 +177,19 @@ in
...
 
         devShellHook = ''
           #  # Find the project root and add the `bin' directory to `PATH'.
           if git rev-parse --is-inside-work-tree >/dev/null 2>&1; then
-            PATH="$( git rev-parse --show-toplevel; )/pkgdb/bin":$PATH;
+            REPO_ROOT="$( git rev-parse --show-toplevel; )";
+            PATH="$REPO_ROOT/pkgdb/bin:$PATH";
+            PKGDB_BIN="$REPO_ROOT/pkgdb/bin/pkgdb";
+            PKGDB_SEARCH_PARAMS_BIN="$REPO_ROOT/pkgdb/tests/search-params";
+            PKGDB_IS_SQLITE3_BIN="$REPO_ROOT/pkgdb/tests/is_sqlite3";
```
Comment (aakropotkin) [inline pkgs/flox-pkgdb/default.nix:192]: @garbas possibly relevant

File: pkgdb/Makefile [hunk 5]
```diff
@@ -145,7 +145,7 @@ endif  # ifndef TOOLCHAIN
 # Read info from files
 # --------------------
 
-VERSION := $(file < $(PKGDB_ROOT)/version)
```
Comment (aakropotkin) [inline pkgdb/Makefile:148]: This avoids an issue where `#include <version>` inside of `boost` can wind up including this file if you are using a non-GNU version of `make`.

**Review by ysndr — CHANGES_REQUESTED** [review]

File: pkgdb/Makefile [hunk 6]
```diff
@@ -520,6 +520,28 @@ lint: compile_commands.json $(HEADERS) $(ALL_SRCS)
 	$(TIDY) $(filter-out compile_commands.json,$^);
 
 
+# ---------------------------------------------------------------------------- #
+
+<<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (aakropotkin) [reply pkgdb/Makefile:525]: good catch

File: pkgdb/src/pkgdb/scrape.cc [hunk 7]
```diff
@@ -27,6 +32,18 @@ ScrapeCommand::ScrapeCommand() : parser( "scrape" )
...
     .nargs( 0 )
     .action( [&]( const auto & ) { this->force = true; } );
+  // TODO: don't assign to global.
+  //       Extend `PkgDbInput' and pass this through the registry as a member of
+  //       each individual flake input.
+  //       For now since we only have one input we'll use this forall.
+  //       This needs a rework later.
+  //
+  // TODO: Using this flag should invalidate the cache.
+  // TODO: We need to invalidate caches any time we change the rules file.
```
[the remark this answers is not in the document]
Comment (aakropotkin) [reply pkgdb/src/pkgdb/scrape.cc:42]: yes

File: pkgdb/include/flox/pkgdb/write.hh [hunk 8]
```diff
@@ -27,6 +31,107 @@ using Target = std::tuple<flox::AttrPath, flox::Cursor, row_id>;
...
+  SR_DISALLOW_RECURSIVE
+}; /* End enum `ScrapeRule` */
+
+[[nodiscard]] std::string
+scrapeRuleToString( ScrapeRule rule );
+
+
+/* -------------------------------------------------------------------------- */
+
+struct RulesTreeNode
```
[the remark this answers is not in the document]
Comment (aakropotkin) [reply pkgdb/include/flox/pkgdb/write.hh:70]: Yeah that's a good point I can get some docs added.

File: pkgdb/src/pkgdb/rules.json.hh
[the remark this answers is not in the document]
Comment (aakropotkin) [reply pkgdb/src/pkgdb/rules.json.hh:1]: Yes these are the defaults.

File: pkgdb/src/pkgdb/write.cc [hunk 9]
```diff
@@ -7,19 +7,295 @@
...
+  const RulesTreeNode * node = this;
+  for ( const auto & attrName : path )
+    {
+      try
+        {
+          node = &node->children.at( attrName );
+        }
+      catch ( const std::out_of_range & err )
+        {
+          return SR_DEFAULT;
```
[the remark this answers is not in the document]
Comment (aakropotkin) [reply pkgdb/src/pkgdb/write.cc:131]: `SR_DEFAULT` means "run the regular `scrape` checks, but nothing special" - so check `recurseForDerivations` and junk.

`SR_UNSET` is an uninitialized state.

File: pkgdb/src/pkgdb/write.cc [hunk 10]
```diff
@@ -7,19 +7,295 @@
...
+  switch ( rule )
+    {
+      case SR_ALLOW_PACKAGE: return true;
+      case SR_DISALLOW_PACKAGE: return false;
+      case SR_ALLOW_RECURSIVE: return true;
+      case SR_DISALLOW_RECURSIVE: return false;
+      case SR_DEFAULT: return std::nullopt;
+      default:
+        throw PkgDbException( "encountered unexpected rule `"
+                              + scrapeRuleToString( rule ) + '\'' );
```
[the remark this answers is not in the document]
Comment (aakropotkin) [reply pkgdb/src/pkgdb/write.cc:164]: I have always been using ``some doc `keyword'``; it's only in follow up PRs that they're getting changed to `'keyword'` instead.

That convention for quotations is from LISP and most GNU documentation which is where I picked it up.

File: pkgdb/src/resolver/environment.cc [hunk 11]
```diff
@@ -573,7 +573,7 @@ Environment::tryResolveGroup( const GroupName &          name,
 
   std::vector<std::string> ids;
   for ( const auto & [id, _] : group ) { ids.emplace_back( id ); }
-  std::string groupStr = joinWithDelim( ids, " " );
+  std::string groupStr = concatStringsSep( " ", ids );
```
[the remark this answers is not in the document]
Comment (aakropotkin) [reply pkgdb/src/resolver/environment.cc:576]: I pulled the definition from `nix` and added it to our `util.hh` file.
Then aligned all invocations of `nix::concatStringsSep( ... )` and `joinWithDelim( ... )` to use _our_ `concatStringsSep( ... )`.

File: pkgdb/src/pkgdb/write.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (ysndr) [reply pkgdb/src/pkgdb/write.cc:131]: is `SR_UNSET` then an error state or handled like default or just ignored basically?

**Comment by samrose** [comment]
This is an integration test case meant to be used with scrape rules. It probably doesn't work as-is but is a good start


```shell
#! /usr/bin/env bats
# -*- mode: bats; -*-
# ============================================================================ #
#
# `pkgdb scrape --rules' CLI tests.
#
# These tests are meant to test the `pkgdb scrape --rules' CLI command.
#
# ---------------------------------------------------------------------------- #

load setup_suite.bash

# bats file_tags=cli,scrape,scrape-rules,flake

# Set up test environment
setup() {
	export DBPATH="$BATS_FILE_TMPDIR/test.sqlite"
	mkdir -p "$BATS_FILE_TMPDIR"
	# We don't parallelize these to avoid DB sync headaches and to recycle the
	# cache between tests.
	# Nonetheless this file makes an effort to avoid depending on past state in
	# such a way that would make it difficult to eventually parallelize in
	# the future.
	export BATS_NO_PARALLELIZE_WITHIN_FILE=true

	# Define rules.json content
	rules_json='{
				"allowRecursive": [
					[
						"legacyPackages",
						null,
						"darwin"
					]
				],
				"allowPackage": [
					[
						"legacyPackages",
						null,
						"python310Packages",
						"pip"
					],
					[
						"legacyPackages",
						null,
						"python310Packages",
						"django"
					]
				],
				"disallowRecursive": [
					[
						"legacyPackages",
						null,
						"python310Packages"
					],
					[
						"legacyPackages",
						null,
						"emacsPackages"
					],
					[
						"legacyPackages",
						null,
						"vimPlugins"
					]
				],
				"disallowPackage": [
					[
						"legacyPackages",
						null,
						"sqlite"
					]
				]
			}'
} #end setup

# Test pkgdb scrape with allow_json
@test "pkgdb scrape with allow_json" {
	#disable for now until feature is implemented
	skip "FIXME: disable for now until feature is implemented"
	run pkgdb scrape <(echo "$rules_json") <NIXPKGS >--rules
	assert_success
	run sqlite3 "$DBPATH" "SELECT attrName FROM Packages      \
    WHERE name = 'legacyPackages.x86_64-linux.darwin' LIMIT 1"
	assert_output 'darwin'
}

@test "pkgdb scrape with disallow emacs" {
	#disable for now until feature is implemented
	skip "FIXME: disable for now until feature is implemented"
	run pkgdb scrape <(echo "$rules_json") <NIXPKGS >--rules
	run sqlite3 "$DBPATH" "SELECT attrName FROM Packages      \
    WHERE name = 'legacyPackages.x86_64-linux.emacs' LIMIT 1"
	assert_output ''
}

@test "pkgdb scrape with both allow_json and disallow_json" {
	#disable for now until feature is implemented
	skip "FIXME: disable for now until feature is implemented"

	run pkgdb scrape <(echo "$rules_json") <NIXPKGS >--rules

	#allowRecursive test
	run sqlite3 "$DBPATH" "SELECT attrName FROM Packages      \
    WHERE name = 'legacyPackages.x86_64-linux.darwin' LIMIT 1"
	assert_output 'darwin'
	#disallowRecursive test
	run sqlite3 "$DBPATH" "SELECT attrName FROM Packages      \
    WHERE name = 'legacyPackages.x86_64-linux.emacs' LIMIT 1"
	assert_output ''
	#requests should not be found due to disallowRecursive
	run sqlite3 "$DBPATH" "SELECT attrName FROM Packages      \
    WHERE name = 'legacyPackages.x86_64-linux.python3Packages.requests' LIMIT 1"
	assert_output ''
	#pip and django should be found allowPackage takes priority over disallowRecursive.
	# that lets us "disable all python310Packages except for ____ (pip and django)"
	run sqlite3 "$DBPATH" "SELECT attrName FROM Packages      \
    WHERE name = 'legacyPackages.x86_64-linux.python3Packages.pip' LIMIT 1"
	assert_output 'pip'
	run sqlite3 "$DBPATH" "SELECT attrName FROM Packages      \
    WHERE name = 'legacyPackages.x86_64-linux.python3Packages.django' LIMIT 1"
	assert_output 'django'
	#disallowPackage test
	run sqlite3 "$DBPATH" "SELECT attrName FROM Packages      \
    WHERE name = 'legacyPackages.x86_64-linux.sqlite' LIMIT 1"
	assert_output ''
}
```

**Comment by billlevine** [comment]
Superseded by PR #1095 .

**Pull request closed without merging**